[RFC] vim-patch: 8.0.0198, 8.0.0200, 8.0.0201, 8.0.0202, 8.0.0204

dvejmz opened the pull request:
Problem:    Some syntax arguments take effect even after "if 0". (Taylor
            Venable)
Solution:   Properly skip the syntax statements.  Make "syn case" and "syn
            conceal" report the current state.  Fix that "syn clear" didn't
            reset the conceal flag.  Add tests for :syntax skipping properly.

https://github.com/vim/vim/commit/de318c5c35ed0d65fd2a07196cb8acd5ee6d9bf8

First time merging a Vim patch (or working on Neovim's codebase for that matter), my apologies if I didn't follow the protocol to the letter. I'm quite happy to change anything if necessary.

I've made some very minor code style changes on this PR (conditional statement curly brace positioning mostly) so this patch satisfies Neovim's coding standards.

Commit:
vim-patch:8.0.0198

Problem:    Some syntax arguments take effect even after "if 0". (Taylor
            Venable)
Solution:   Properly skip the syntax statements.  Make "syn case" and "syn
            conceal" report the current state.  Fix that "syn clear" didn't
            reset the conceal flag.  Add tests for :syntax skipping properly.

https://github.com/vim/vim/commit/de318c5c35ed0d65fd2a07196cb8acd5ee6d9bf8

Shougo commented:
Related patches: 0200, 0201, 0202, 0204

justinmk commented:
Thanks! A `src/nvim/testdir/...` test is failing now (see Travis; these tests don't run on QuickBuild/AppVeyor):

```
From test_syntax.vim:
Found errors in Test_syntax_arg_skipped():
function RunTheTest[13]..Test_syntax_arg_skipped line 22: Pattern 'conceal off' does not match '\nsyn conceal on'
TEST FAILURE
```

You can [run that test locally](https://github.com/neovim/neovim/wiki/Building-Neovim#running-tests) like this:

    TEST_FILE=test_syntax.res make oldtest

dvejmz commented:
Thanks for letting me know @justinmk . That is the new test that I added with this patch. I ran the new test suite (unit and functional) but didn't do the same with the old suite. I'll fix that.

I'll also go back and merge the related patches @Shougo pointed out.

Marking this as WIP.

jamessan commented:
@dvejmz Do you think you'll have time to work on this again?

dvejmz commented:
@jamessan Yes! I do apologise, got sidetracked by various other things and never got around working on this again. I'll make sure to find some time this week to finish this PR off or at least make good progress with it.

Commit:
vim-patch:8.0.0200: some syntax arguments are not tested

Problem:    Some syntax arguments are not tested.
Solution:   Add more syntax command tests.

https://github.com/vim/vim/commit/58f60ca2fcd2858faac84e386b3ccf5ced75084d

Commit:
vim-patch:8.0.0201: completion of highlight groups includes cleared names

Problem:    When completing a group name for a highlight or syntax command
            cleared groups are included.
Solution:   Skip groups that have been cleared.

https://github.com/vim/vim/commit/d61e8aaae57bd66279def479462bf11c22ec2f1c

Commit:
vim-patch:8.0.0202: no test for invalid syntax group name

Problem:    No test for invalid syntax group name.
Solution:   Add a test for group name error and warning.

https://github.com/vim/vim/commit/4007ed4a5e8c34197078e9d5718bd1d4a429dd23

Commit:
vim-patch:8.0.0204: compiler warns for uninitialized variable

Problem:    Compiler warns for uninitialized variable. (Tony Mechelynck)
Solution:   When skipping set "id" to -1.

https://github.com/vim/vim/commit/eb46f8fa14a586779f55b1c7f1648f559618322e

dvejmz commented:
Firstly, my apologies for the massive delay in the delivery of this PR! Anyhow, all related patches have now been ported over and the old test suite is all green. New test suite fails in the same places as in master so,

I've removed all `'#define`s and `has()` checks in the test file for toggled Vim features, as they're all included by default on Neovim.

This should be ready to review now. Thanks.

justinmk commented:
> I've removed all #defines and has() checks in the test file for toggled Vim features, as they're unnecessary,

Old test should stay identical if possible, makes merges easier.

justinmk commented:
Also please check `make lint` and include lint fixes in a separate "lint" commit.

Commit:
lint

dvejmz commented:
Addressed last two feedback items @justinmk 
* Added back feature check to test
* Fixed all lint errors

dvejmz commented on src/nvim/syntax.c:
```diff
@@ -7730,10 +7773,20 @@ const char *get_highlight_name(expand_T *const xp, const int idx)
+  }
+
+  // Items are never removed from the table, skip the ones that were cleared.
+  int current_idx = idx;
```
The only thing I want to highlight from this PR is this temp variable here, which has to be used in place of `idx` as the code tries to increment `idx` but `idx` is now passed as a `const int` to the function.

justinmk replied:
That makes sense and I think it's good as you have it. Thanks for mentioning.
